### PR TITLE
Add Service type in an exception message from ServiceRegistry

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR adds the Service type as part of the exception message when a ServiceRegistry has not been set for the current thread. This information could help debugging semi-rare failures in HLT tests that were reported in https://github.com/cms-sw/cmssw/pull/38801#issuecomment-1267976632 .

#### PR validation:

Code compiles

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: FWCore/ServiceRegistry/interface/ServiceRegistry.h [hunk 1]
```diff
@@ -61,9 +61,10 @@ namespace edm {
     template <typename T>
     T& get() const {
       if (nullptr == manager_.get()) {
-        Exception::throwThis(errors::NotFound,
-                             "Service"
-                             " no ServiceRegistry has been set for this thread");
+        Exception::throwThis(
+            errors::NotFound,
+            "No ServiceRegistry has been set for this thread. Tried to get a Service with compiler type name ",
+            typeid(T).name());
```
Comment (fwyzard) [inline FWCore/ServiceRegistry/interface/ServiceRegistry.h:67]: depending what they look like, would it make sense to demangle these ?
```suggestion
            boost::core::demangle(typeid(T).name()));
```

**Comment by makortel** [comment]
@cmsbuild, please abort

File: FWCore/ServiceRegistry/interface/ServiceRegistry.h (hunk 1 above)
Comment (makortel) [reply FWCore/ServiceRegistry/interface/ServiceRegistry.h]: Thanks for the suggestion. While this particular exception points to a framework bug, there is another one in `ServicesManager.h` that is thrown if a requested Service is missing from the configuration. What I recall from actually seeing that exception, the mangled name has usually(?) been informative enough to point to right direction, but a demangled name would certainly be more clear. I changed the two exceptions here as well as the one in `ServicesManager` to use demangled name.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
The unit test failures are unrelated. The RelVals-INPUT failure looks like a job was killed?

For AddOn tests the link to `hlt_mc_HIon` is wrong (@smuzaffar), the failed test was actually `hlt_data_HIon`. This was actually lucky because now we see the effect of this PR
```
----- Begin Fatal Exception 06-Oct-2022 12:34:22 CEST-----------------------
An exception of category 'NotFound' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
No ServiceRegistry has been set for this thread. Tried to ask availability of a Service edm::SiteLocalConfig
----- End Fatal Exception -------------------------------------------------
```
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0130f6/28033/addOnTests/hlt_data_HIon/cmsRun__cvmfs_cms-ib.cern.ch_nweek-02753_el8_amd64_gcc10_cms_cmssw-patch_CMSSW_12_6_X_2022-10-05-1100_src_HLTrigger_Configuration_test_OnLine_HLT_HIon.log

(further confirmation of https://github.com/cms-sw/cmssw/pull/38801#issuecomment-1268885816)

**Comment by makortel** [comment]
+1

**Comment by rappoccio** [comment]
Hoping the failed tests are spurious, will try again.

**Comment by rappoccio** [comment]
please test

**Comment by smuzaffar** [comment]
> For AddOn tests the link to `hlt_mc_HIon` is wrong (@smuzaffar), the failed test was actually `hlt_data_HIon`. This was actually lucky because now we see the effect of this PR

thanks for catching this. The reason behind this wrong like is that both `hlt_data_HIon` and `hlt_mc_HIon` use same logfile names ( [first 150 characters of the command](https://github.com/cms-sw/cmssw/blob/master/Utilities/ReleaseScripts/scripts/addOnTests.py#L40) )
```
hlt_data_HIon/cmsRun__cvmfs_cms-ib.cern.ch_nweek-02753_el8_amd64_gcc10_cms_cmssw-patch_CMSSW_12_6_X_2022-10-05-1100_src_HLTrigger_Configuration_test_OnLine_HLT_HIon.log
hlt_mc_HIon/cmsRun__cvmfs_cms-ib.cern.ch_nweek-02753_el8_amd64_gcc10_cms_cmssw-patch_CMSSW_12_6_X_2022-10-05-1100_src_HLTrigger_Configuration_test_OnLine_HLT_HIon.log
```

which confuses bot

**Comment by missirol** [comment]
> For AddOn tests the link to `hlt_mc_HIon` is wrong

For the record, this has happened before (https://github.com/cms-sw/cmssw/pull/36408#issuecomment-988888689). https://github.com/cms-sw/cmssw/pull/36423 was an attempt to fix it, but it was not integrated (while https://github.com/cms-sw/cms-bot/pull/1675 was..).

**Comment by perrotta** [comment]
There are DQM differences for JetMET in several workflows: not clear if and why this PR can have generated them...

**Comment by fwyzard** [comment]
I doubt this PR can cause any changes in physics or reconstruction quantities...

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- The differences in the DQM comparisons have disappeared
- The unit test error didn't show up in previous tests, and t is clearly unrelated

**Comment by perrotta** [comment]
merge

**Pull request merged**